diff -r 4617188a5091 me/blocks/vnic/shared/nfd_cfg_internal.c
--- a/me/blocks/vnic/shared/nfd_cfg_internal.c	Wed Jul 19 19:45:30 2017 -0400
+++ b/me/blocks/vnic/shared/nfd_cfg_internal.c	Tue Aug 22 16:52:58 2017 -0400
@@ -568,11 +568,14 @@
 #ifndef NFD_BPF_MAX_LEN
 #define NFD_BPF_MAX_LEN (8 * 1024 - NFD_BPF_START_OFF)
 #endif
+#ifndef NFD_BPF_STACK_SZ
+#define NFD_BPF_STACKSZ 0
+#endif
 
     __xwrite unsigned int bpf_cfg[] =
         { (NFD_BPF_ABI | (NFD_BPF_CAPS << 8) | (NFD_BPF_MAX_LEN << 16)),
           NFD_BPF_START_OFF | NFD_BPF_DONE_OFF << 16,
-          30 << 8 /* CTM buf size / 64 */ };
+          NFD_BPF_STACK_SZ / 64 | 30 << 8 /* CTM buf size / 64 */ };
 #endif
 
     mem_write64(&cfg, NFD_CFG_BAR_ISL(PCIE_ISL, vid) + NFP_NET_CFG_VERSION,
diff -r 4617188a5091 me/blocks/vnic/shared/nfd_flr.c
--- a/me/blocks/vnic/shared/nfd_flr.c	Wed Jul 19 19:45:30 2017 -0400
+++ b/me/blocks/vnic/shared/nfd_flr.c	Tue Aug 22 16:52:58 2017 -0400
@@ -186,11 +186,14 @@
 #ifndef NFD_BPF_MAX_LEN
 #define NFD_BPF_MAX_LEN (8 * 1024 - NFD_BPF_START_OFF)
 #endif
+#ifndef NFD_BPF_STACK_SZ
+#define NFD_BPF_STACK_SZ 0
+#endif
 
     __xwrite unsigned int bpf_cfg[] =
         { (NFD_BPF_ABI | (NFD_BPF_CAPS << 8) | (NFD_BPF_MAX_LEN << 16)),
           NFD_BPF_START_OFF | NFD_BPF_DONE_OFF << 16,
-          30 << 8 /* CTM buf size / 64 */ };
+          NFD_BPF_STACK_SZ / 64 | 30 << 8 /* CTM buf size / 64 */ };
 #endif
 
     mem_write64(&cfg,
