diff -r cbad0447831e kernel/nfp_net_ctrl.h
--- a/kernel/nfp_net_ctrl.h	Mon May 15 22:08:40 2017 +0200
+++ b/kernel/nfp_net_ctrl.h	Wed Jul 05 13:25:38 2017 -0400
@@ -118,7 +118,7 @@
 					 NFP_NET_CFG_CTRL_LSO2)
 #define   NFP_NET_CFG_CTRL_RSS_ANY	(NFP_NET_CFG_CTRL_RSS | \
 					 NFP_NET_CFG_CTRL_RSS2)
-#define   NFP_NET_CFG_CTRL_CHAIN_META	(NFP_NET_CFG_CTRL_RSS2 \
+#define   NFP_NET_CFG_CTRL_CHAIN_META	(NFP_NET_CFG_CTRL_RSS2 | \
 					 NFP_NET_CFG_CTRL_CSUM_COMPLETE)
 
 #define NFP_NET_CFG_UPDATE              0x0004
