diff -r 98afbcd74ef6 me/blocks/vnic/svc/msix_qmon.c
--- a/me/blocks/vnic/svc/msix_qmon.c	Wed Apr 05 14:23:10 2017 +0200
+++ b/me/blocks/vnic/svc/msix_qmon.c	Mon May 01 13:18:23 2017 -0700
@@ -658,7 +658,7 @@
     queue = NFD_NATQ2BMQ(queue);
 
     addr_hi = (0x84 | (pcie_isl + 4)) << 24;
-    addr_lo = queue * NFD_OUT_ATOMICS_SZ + NFD_OUT_ATOMICS_SENT;
+    addr_lo = queue * NFD_OUT_ATOMICS_SZ + NFD_OUT_ATOMICS_DMA_DONE;
 
     __asm mem[atomic_read, rdata, addr_hi, <<8, addr_lo, 1], ctx_swap[rsig];
 
