diff -r 64ea5722b61e me/blocks/vnic/shared/nfd_cfg_internal.c
--- a/me/blocks/vnic/shared/nfd_cfg_internal.c	Mon Jul 10 20:13:16 2017 -0400
+++ b/me/blocks/vnic/shared/nfd_cfg_internal.c	Mon Jul 10 20:25:45 2017 -0400
@@ -562,9 +562,12 @@
 #ifndef NFD_BPF_ABI
 #define NFD_BPF_ABI (NFP_NET_BPF_ABI)
 #endif
+#ifndef NFD_BPF_CAPS
+#define NFD_BFP_CAPS 0
+#endif
 
     __xwrite unsigned int bpf_cfg[] =
-        { NFD_BPF_ABI | (8 * 1024 - NFD_BPF_START_OFF) << 16,
+        { NFD_BPF_ABI | (NFD_BPF_CAPS << 8) | (8 * 1024 - NFD_BPF_START_OFF) << 16,
           NFD_BPF_START_OFF | NFD_BPF_DONE_OFF << 16,
           30 << 8 /* CTM buf size / 64 */ };
 #endif
diff -r 64ea5722b61e me/blocks/vnic/shared/nfd_flr.c
--- a/me/blocks/vnic/shared/nfd_flr.c	Mon Jul 10 20:13:16 2017 -0400
+++ b/me/blocks/vnic/shared/nfd_flr.c	Mon Jul 10 20:25:45 2017 -0400
@@ -180,8 +180,11 @@
 #ifndef NFD_BPF_ABI
 #define NFD_BPF_ABI (NFP_NET_BPF_ABI)
 #endif
+#ifndef NFD_BPF_CAPS
+#define NFD_BFP_CAPS 0
+#endif
 
-    __xwrite unsigned int bpf_cfg[] = { NFD_BPF_ABI | (8 * 1024 - NFD_BPF_START_OFF) << 16,
+    __xwrite unsigned int bpf_cfg[] = { NFD_BPF_ABI | (NFD_BPF_CAPS << 8) | (8 * 1024 - NFD_BPF_START_OFF) << 16,
                                         NFD_BPF_START_OFF | NFD_BPF_DONE_OFF << 16,
                                         30 << 8 /* CTM buf size / 64 */ };
 #endif
