diff -r cbad0447831e me/blocks/vnic/shared/nfd_cfg_internal.c
--- a/me/blocks/vnic/shared/nfd_cfg_internal.c	Mon May 15 22:08:40 2017 +0200
+++ b/me/blocks/vnic/shared/nfd_cfg_internal.c	Thu Jun 29 18:57:40 2017 -0400
@@ -453,8 +453,12 @@
     __xwrite unsigned int cfg2[] = {NFD_OUT_RX_OFFSET,
                                     NFD_RSS_HASH_FUNC};
 #ifdef NFD_BPF_CAPABLE
+#ifndef NFD_BPF_CAPS
+#define NFD_BFP_CAPS 0
+#endif
+
     __xwrite unsigned int bpf_cfg[] =
-        { NFP_NET_BPF_ABI | (8 * 1024 - NFD_BPF_START_OFF) << 16,
+        { NFP_NET_BPF_ABI | (NFD_BPF_CAPS << 8) | (8 * 1024 - NFD_BPF_START_OFF) << 16,
           NFD_BPF_START_OFF | NFD_BPF_DONE_OFF << 16,
           30 << 8 /* CTM buf size / 64 */ };
 #endif
