diff --git a/kernel/nfp_net_ctrl.h b/kernel/nfp_net_ctrl.h
index 6106fd8..fc78dea 100755
--- a/kernel/nfp_net_ctrl.h
+++ b/kernel/nfp_net_ctrl.h
@@ -464,11 +464,29 @@
  * %NFP_NET_CFG_TLV_TYPE_MBOX:
  * Variable, mailbox area.  Overwrites the default location which is
  * %NFP_NET_CFG_MBOX_BASE and length %NFP_NET_CFG_MBOX_VAL_MAX_SZ.
+ *
+ * @NFP_NET_CFG_TLV_TYPE_STATS_BASE_MASK:
+ * Mask of enabled base stats.  Mask has bits set to 1 for @NFP_NET_CFG_STATS_*
+ * stats which are populated by the application FW.  Must be unique.
+ *
+ * @NFP_NET_CFG_TLV_TYPE_EXT_STATS:
+ * Extended statistics.  Multiple instances allowed.
+ * Size:  |  4B   | count * 1B | count * 32B |    count * 8B    |
+ * Field: | count |    type    |     name    | 64bit statistics |
  */
 #define NFP_NET_CFG_TLV_TYPE_UNKNOWN		0
 #define NFP_NET_CFG_TLV_TYPE_RESERVED		1
 #define NFP_NET_CFG_TLV_TYPE_END		2
 #define NFP_NET_CFG_TLV_TYPE_ME_FREQ		3
 #define NFP_NET_CFG_TLV_TYPE_MBOX		4
+#define NFP_NET_CFG_TLV_TYPE_STATS_BASE_MASK    5
+#define NFP_NET_CFG_TLV_TYPE_EXT_STATS          6
+
+#define NFP_NET_EXT_STATS_TYPE_UNKNOWN          0
+#define NFP_NET_EXT_STATS_TYPE_NODIR            1
+#define NFP_NET_EXT_STATS_TYPE_RX               2
+#define NFP_NET_EXT_STATS_TYPE_TX               3
+
+#define NFP_NET_EXT_STATS_NAME_LEN              32
 
 #endif /* _NFP_NET_CTRL_H_ */
