diff --git a/me/blocks/vnic/pci_out_pd.uc b/me/blocks/vnic/pci_out_pd.uc
index a5969164..d4ff7616 100644
--- a/me/blocks/vnic/pci_out_pd.uc
+++ b/me/blocks/vnic/pci_out_pd.uc
@@ -843,7 +843,9 @@ add_wq_credits#:
 
     wsm_extract(isl, io_work, SB_WQ_CTM_ISL)
 #ifndef NFD_OUT_ALWAYS_HAS_CTM
+#ifndef NFD_OUT_ALWAYS_FREE_CTM
     beq[no_ctm_buffer#], defer[1]
+#endif
 #endif
     alu[ring_num, SB_WQ_BLS_msk, AND, io_work[SB_WQ_BLS_wrd], >>SB_WQ_BLS_shf]
 
