diff -r 98afbcd74ef6 me/blocks/vnic/nfd_in.uc
--- a/me/blocks/vnic/nfd_in.uc	Wed Apr 05 14:23:10 2017 +0200
+++ b/me/blocks/vnic/nfd_in.uc	Thu Apr 20 16:56:32 2017 +0200
@@ -152,7 +152,8 @@
 #define NFD_IN_VLAN_fld         3, 15, 0
 #define NFD_IN_META_SIZE        16
 #define NFD_IN_META_SIZE_LW     4
-
+#define NFD_IN_LSO2_L3_OFFS_fld      3, 7, 0
+#define NFD_IN_LSO2_L4_OFFS_fld      3, 15, 8
 
 #define NFD_IN_RING_HI_fld      0, 31, 24
 #define NFD_IN_RING_SP_fld      0, 23, 12
