diff -r 98afbcd74ef6 kernel/nfp_net_ctrl.h
--- a/kernel/nfp_net_ctrl.h	Wed Apr 05 14:23:10 2017 +0200
+++ b/kernel/nfp_net_ctrl.h	Thu Apr 20 15:25:08 2017 +0200
@@ -101,6 +101,7 @@
 #define   NFP_NET_CFG_CTRL_VXLAN	  (0x1 << 24) /* VXLAN tunnel support */
 #define   NFP_NET_CFG_CTRL_NVGRE	  (0x1 << 25) /* NVGRE tunnel support */
 #define   NFP_NET_CFG_CTRL_MSIX_TX_OFF	  (0x1 << 26) /* MSIX TX off */
+#define   NFP_NET_CFG_CTRL_LSO2  	    (0x1 << 28) /* LSO/TSO (version 2) */
 #define NFP_NET_CFG_UPDATE              0x0004
 #define   NFP_NET_CFG_UPDATE_GEN          (0x1 <<  0) /* General update */
 #define   NFP_NET_CFG_UPDATE_RING         (0x1 <<  1) /* Ring config change */
